config/operation.go

```diff
@@ -0,0 +1,496 @@
+	if err != nil {
+		return err
+	}
+
+	u.oauthCreds = token
+	return nil
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+func GenerateAccountKey(c *Config, s *Service, a *Account) (string, error) {
+	j := struct {
+		WebGatewayUrl string
+		Protocol      string
+		PrivateKey    string
+	}{
+		WebGatewayUrl: c.Url,
+		Protocol:      a.ClientCreds.Protocol,
+		PrivateKey:    a.ClientCreds.PrivateKey,
+	}
+
+	b, err := json.Marshal(j)
+	if err != nil {
+		return "", err
+	}
+
+	inner := base64.StdEncoding.EncodeToString(b)
+
+	return fmt.Sprintf("KEYBEGIN_%s/%s_%s_KEYEND", s.ServiceName, a.AccountName, inner), nil
+}
+
+func GenerateOauthConfig(url string, s *Service) (*oauth2.Config, error) {
+	var endpoint = oauth2.Endpoint{
+		AuthURL:  s.OauthServiceCreds.AuthURL,
+		TokenURL: s.OauthServiceCreds.TokenURL,
+	}
+	redirectUrl, err := getRedirectUrl(url)
+	if err != nil {
+		return nil, err
+	}
+	oauthConf := &oauth2.Config{
+		ClientID:     s.OauthServiceCreds.ClientID,
+		ClientSecret: s.OauthServiceCreds.ClientSecret,
+		Scopes:       s.OauthServiceCreds.Scopes,
+		Endpoint:     endpoint,
+		RedirectURL:  redirectUrl,
+	}
+	return oauthConf, nil
+}
+
+func GenerateAuthUrl(oauthConf *oauth2.Config) (string, string, error) {
+	state, err := generateRandomString()
+	if err != nil {
+		return "", "", fmt.Errorf("Problem with random number generation.  Can't continue.\n")
+	}
+	return oauthConf.AuthCodeURL(state), state, nil
+}
+
+func VerifyState(code string, state string) (string, error) {
+	jsonAuthCode, err := base64.StdEncoding.DecodeString(code)
+	if err != nil {
+		return "", fmt.Errorf("Bad decode.\n")
+	}
+	j := struct {
+		Token string
+		State string
+	}{}
+	json.Unmarshal(jsonAuthCode, &j)
+	if j.State != state {
+		return "", fmt.Errorf("Bad state. Expected %s, got %s\n", state, j.State)
+	}
+	return j.Token, nil
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+func verifyName(rawText string) error {
+	if rawText == "" {
+		return errors.New("Name cannot be empty.")
+	}
+
+	match, _ := regexp.MatchString("^[-a-z0-9]+$", rawText)
+	if !match {
+		return errors.New("That name contains invalid characters.")
+	}
+	return nil
+}
+
+func verifyUrl(rawText string) (string, error) {
+	tokenURL, err := url.ParseRequestURI(rawText)
+	if err != nil {
+		return "", errors.New(rawText + " is not a valid URL (include https://)")
+	}
+	if tokenURL.Scheme != "https" {
+		return "", errors.New(rawText + " does not use https.")
+	}
+
+	return tokenURL.String(), nil
+}
+
+func verifyEmail(rawText string) (string, error) {
+	rxEmail := regexp.MustCompile("^[a-zA-Z0-9.!#$%&'*+\\/=?^_`{|}~-]+" +
```

Comment:
regex compilations are usually expensive and it might worth if you take this out and define/compile it only once and use it multiple times here.

Reply:
thanks!
i didnt think about this